### PR TITLE
sensors:am2320: add support for AM2320

### PR DESCRIPTION
**Pull request by chunhuajiang**
Add initinal support for AM2320 Temperature and Humidity sensor.
Tested with frdm_k64f.

Signed-off-by: Tidy Jiang <tidyjiang@163.com>

### CODE REVIEW

File: drivers/sensor/am2320/am2320.c [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+#include <init.h>
+#include <misc/__assert.h>
+#include <misc/byteorder.h>
+#include <sensor.h>
+#include <string.h>
+#include <crc16.h>
+#include <logging/sys_log.h>
+
+/* am2320 i2c address, 7+1 bit */
+#define AM2320_I2C_ADDR     0xB8
```
Comment (bogdan-davidoaia) [inline drivers/sensor/am2320/am2320.c]: Keep this define name for the 7bit address and use a different define for 7+1bit. This way you don't need a shift for each i2c_write.

File: drivers/sensor/am2320/am2320.c [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+	/* Step 0: clean stored old data. */
+	drv_data->temp     = 0;
+	drv_data->humidity = 0;
+
+	/**
+	 * Step 1: Wake up AM2320.
+	 * Send START + ADDRESS(0xA8) signal to wake up AM2320.
```
Comment (bogdan-davidoaia) [inline drivers/sensor/am2320/am2320.c]: The value of the define is 0xB8, why is it 0xA8 here?

File: drivers/sensor/am2320/am2320.c [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	 * We do not check the return value, as AM2320 will not
+	 * response ACK signal.
+	 */
+	buf[0] = AM2320_I2C_ADDR;
+	i2c_write(drv_data->i2c, buf, 1, AM2320_I2C_ADDR >> 1);
+
+	/* Step 2: Send Read Command to AM2320. */
+	buf[0] = 0x03;
```
Comment (bogdan-davidoaia) [inline drivers/sensor/am2320/am2320.c]: Use defines for these values.

File: drivers/sensor/am2320/am2320.c [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	if (crc != am2320_crc16(buf, 6)) {
+		SYS_LOG_ERR("Failed to fetch data sample.");
+		return -EIO;
+	}
+
+	/* Step 5: store data to driver structure. */
+	drv_data->temp     = buf[4] * 256 + buf[5];
+	drv_data->humidity = buf[2] * 256 + buf[3];
+	SYS_LOG_INF("fetch: temp:%d, humitidy:%d\n",
+					drv_data->temp, drv_data->humidity);
```
Comment (bogdan-davidoaia) [inline drivers/sensor/am2320/am2320.c]: Fixa alignment.

File: drivers/sensor/am2320/am2320.c [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+	/* Step: 3: Read data from AM2320. */
+	memset(buf, 0, sizeof(buf));
+	if (i2c_read(drv_data->i2c, buf, 8, AM2320_I2C_ADDR >> 1)) {
+		SYS_LOG_ERR("Failed to fetch data sample.");
+		return -EIO;
+	}
+
+	/* Step 4: verify data.  */
+	crc = buf[7] * 256 + buf[6];
```
Comment (bogdan-davidoaia) [inline drivers/sensor/am2320/am2320.c]: Use << 8 rather then * 256.

File: drivers/sensor/am2320/am2320.c [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+};
+
+int am2320_init(struct device *dev)
+{
+	struct am2320_data *drv_data = dev->driver_data;
+
+	drv_data->i2c = device_get_binding(CONFIG_AM2320_I2C_MASTER_DEV_NAME);
+	if (drv_data->i2c == NULL) {
+		SYS_LOG_ERR("Could not get pointer to %s device.",
+				CONFIG_AM2320_I2C_MASTER_DEV_NAME);
```
Comment (bogdan-davidoaia) [inline drivers/sensor/am2320/am2320.c]: Fix alignment.

File: drivers/sensor/am2320/am2320.c [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+	dev->driver_api = &am2320_driver_api;
+
+	return 0;
+}
+
+struct am2320_data am2320_driver;
+
+DEVICE_INIT(am2320, CONFIG_AM2320_NAME, am2320_init, &am2320_driver,
+		NULL, POST_KERNEL, CONFIG_SENSOR_INIT_PRIORITY);
```
Comment (bogdan-davidoaia) [inline drivers/sensor/am2320/am2320.c]: Fix alignment.

File: drivers/sensor/am2320/am2320.c [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+	drv_data->humidity = 0;
+
+	/**
+	 * Step 1: Wake up AM2320.
+	 * Send START + ADDRESS(0xA8) signal to wake up AM2320.
+	 * We do not check the return value, as AM2320 will not
+	 * response ACK signal.
+	 */
+	buf[0] = AM2320_I2C_ADDR;
+	i2c_write(drv_data->i2c, buf, 1, AM2320_I2C_ADDR >> 1);
```
Comment (bogdan-davidoaia) [inline drivers/sensor/am2320/am2320.c]: Are you sure this is needed? The start + address signal mentioned in the datasheet are included when doing an i2c_write.

File: drivers/sensor/am2320/am2320.c [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+					enum sensor_channel chan,
+					struct sensor_value *val)
+{
+	struct am2320_data *drv_data = dev->driver_data;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_TEMP || SENSOR_CHAN_HUMIDITY);
+
+	if (chan == SENSOR_CHAN_TEMP) {
+		val->val1 = drv_data->temp / 10;
+		val->val2 = drv_data->temp % 10;
```
Comment (bogdan-davidoaia) [inline drivers/sensor/am2320/am2320.c]: val2 should be in micro, so val2 = temp % 10 * 100000

File: drivers/sensor/am2320/am2320.c [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+	} else { /* SENSOR_CHAN_HUMIDITY */
+		val->val1 = drv_data->humidity / 10;
```
Comment (bogdan-davidoaia) [inline drivers/sensor/am2320/am2320.c]: Humidity should be in milli percent, so val1 = humidity * 100 and val2 = 0.

File: drivers/sensor/am2320/am2320.c [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+static int am2320_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct am2320_data *drv_data = dev->driver_data;
+	u8_t   buf[8];
+	u16_t  crc;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+	/* Step 0: clean stored old data. */
```
Comment (bogdan-davidoaia) [inline drivers/sensor/am2320/am2320.c]: No other driver clears the previous sample if a fetch error occurs, so you don't need to do this.

**Review by bogdan-davidoaia — CHANGES_REQUESTED** [review]

File: drivers/sensor/am2320/am2320.c (hunk 8 above)
Comment (tidyjiang8) [inline drivers/sensor/am2320/am2320.c]: Yes, you are right, it is included in i2c_write. But I still can't find a way to wake up AM2320.
I tested with sampes/sensor/hts221 just now, if I uncomment this, application can only successfully fetch data once, and then failed the next time. Do you have any good idea?

Comment (bogdan-davidoaia) [inline drivers/sensor/am2320/am2320.c]: You are right, the datasheet does state that you need to explicitly wake up the chip fist. Initially I though that it was done by the implicit address sent by the i2c_write.

So you can leave the way you had in the first version. Sorry for the confusion.

**Review by bogdan-davidoaia — APPROVED** [review]

File: drivers/sensor/am2320/am2320.c [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+	return crc;
+}
+
+static int am2320_channel_get(struct device *dev,
+				enum sensor_channel chan,
+				struct sensor_value *val)
+{
+	struct am2320_data *drv_data = dev->driver_data;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_TEMP || SENSOR_CHAN_HUMIDITY);
```
Comment (MaureenHelm) [inline drivers/sensor/am2320/am2320.c]: This should be
```c
__ASSERT_NO_MSG((chan == SENSOR_CHAN_TEMP) || (chan == SENSOR_CHAN_HUMIDITY));
```

File: drivers/sensor/am2320/am2320.c [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+	if (chan == SENSOR_CHAN_TEMP) {
+		val->val1 = drv_data->temp / 10;
```
Comment (MaureenHelm) [inline drivers/sensor/am2320/am2320.c]: Please add a comment that describes what units you are converting from and to. Same for humidity

File: drivers/sensor/am2320/am2320.c [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+	}
+
+	dev->driver_api = &am2320_driver_api;
+
+	return 0;
+}
+
+struct am2320_data am2320_driver;
+
+DEVICE_INIT(am2320, CONFIG_AM2320_NAME, am2320_init, &am2320_driver,
```
Comment (MaureenHelm) [inline drivers/sensor/am2320/am2320.c]: use `DEVICE_AND_API_INIT` instead and remove line 143

File: drivers/sensor/am2320/am2320.c [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+
+	return 0;
+}
+
+static const struct sensor_driver_api am2320_driver_api = {
+	.sample_fetch = am2320_sample_fetch,
+	.channel_get  = am2320_channel_get,
+};
+
+int am2320_init(struct device *dev)
```
Comment (MaureenHelm) [inline drivers/sensor/am2320/am2320.c]: this should be static

File: drivers/sensor/am2320/am2320.c [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+			    CONFIG_AM2320_I2C_MASTER_DEV_NAME);
+		return -EINVAL;
+	}
+
+	dev->driver_api = &am2320_driver_api;
+
+	return 0;
+}
+
+struct am2320_data am2320_driver;
```
Comment (MaureenHelm) [inline drivers/sensor/am2320/am2320.c]: this should be static

File: drivers/sensor/am2320/am2320.c [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+#define AM2320_FETCH_START_ADDR 0X00
+#define AM2320_FETCH_LEN        0X04
+
+struct am2320_data {
+	struct device *i2c;
+	s16_t temp;
+	s16_t humidity;
+};
+
+static u16_t am2320_crc16(u8_t *ptr, unsigned int len)
```
Comment (MaureenHelm) [inline drivers/sensor/am2320/am2320.c]: Can you use `include/crc16.h` instead?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please add a build test for this driver in another commit. Probably can add to `tests/drivers/build_all/sensors_a_m.conf`

**Comment by nashif** [comment]
outdated, closing. Reopen with updates/rebase if needed.

**Pull request closed without merging**